fl4health/clients/apfl_client.py

```diff
@@ -1,255 +1,101 @@
+        for the local model and one optimizer for the global model.
+        """
+        global_optimizer.param_groups.clear()
+        global_optimizer.state.clear()
+        local_optimizer = copy.deepcopy(global_optimizer)
+
+        global_optimizer.add_param_group({"params": [p for p in self.model.global_model.parameters()]})
+        local_optimizer.add_param_group({"params": [p for p in self.model.local_model.parameters()]})
+        return global_optimizer, local_optimizer
+
+    def setup_client(self, config: Config) -> None:
+        """
+        Set dataloaders, optimizers, parameter exchangers and other attributes derived from these.
+        """
+        super().setup_client(config)
+
+        # Split optimizer from get_optimizer into two distinct optimizers
+        # One for local model and one for global model
+        global_optimizer, local_optimizer = self.split_optimizer(self.optimizer)
+        self.optimizer = global_optimizer
+        self.local_optimizer = local_optimizer
+
+    def train_step(self, input: torch.Tensor, target: torch.Tensor) -> Tuple[Losses, torch.Tensor]:
         # Mechanics of training loop follow from original implementation
         # https://github.com/MLOPTPSU/FedTorch/blob/main/fedtorch/comms/trainings/federated/apfl.py
 
         # Forward pass on global model and update global parameters
-        self.global_optimizer.zero_grad()
-        global_pred = self.model(input, personal=False)["global"]
+        self.optimizer.zero_grad()
+        global_pred = self.model.global_forward(input)
         global_loss = self.criterion(global_pred, target)
         global_loss.backward()
-        self.global_optimizer.step()
+        self.optimizer.step()
 
         # Make sure gradients are zero prior to foward passes of global and local model
         # to generate personalized predictions
         # NOTE: We zero the global optimizer grads because they are used (after the backward calculation below)
         # to update the scalar alpha (see update_alpha() where .grad is called.)
-        self.global_optimizer.zero_grad()
+        self.optimizer.zero_grad()
         self.local_optimizer.zero_grad()
 
         # Personal predictions are generated as a convex combination of the output
         # of local and global models
-        pred_dict = self.model(input, personal=True)
-        personal_pred, local_pred = pred_dict["personal"], pred_dict["local"]
+        personal_pred = self.predict(input)
 
         # Parameters of local model are updated to minimize loss of personalized model
-        personal_loss = self.criterion(personal_pred, target)
-        personal_loss.backward()
+        losses = self.compute_loss(personal_pred, target)
+        losses.backward.backward()
         self.local_optimizer.step()
 
-        with torch.no_grad():
-            local_loss = self.criterion(local_pred, target)
-
-        return local_loss, global_loss, personal_loss, local_pred, global_pred, personal_pred
```

Comment:
Could we not fold the local and global losses into the Losses class as auxiliary losses as done with the proximal term for FedProx? I found these losses to be quite useful in debugging and understanding the training of APFL for the FLamby experiments.

Reply:
That is, defining a custom `compute_loss` as done for FedProx?

Reply:
Updated in https://github.com/VectorInstitute/FL4Health/pull/62/commits/f462d69c915f7f19a2d97d808e685d1d0ad39dd6